src/compute-types/src/time_dependence.rs

```diff
@@ -0,0 +1,220 @@
+use proptest::strategy::Strategy;
+use serde::{Deserialize, Serialize};
+
+include!(concat!(
+    env!("OUT_DIR"),
+    "/mz_compute_types.time_dependence.rs"
+));
+
+/// Description of how a dataflow follows time.
+///
+/// The default value indicates the dataflow follows wall-clock without modifications.
+#[derive(Debug, Clone, Default, Serialize, Deserialize, Eq, PartialEq, Ord, PartialOrd)]
+pub struct TimeDependence {
+    /// Optional refresh schedule. None indicates no rounding.
+    pub schedule: Option<RefreshSchedule>,
+    /// Inner dependencies to evaluate first. Empty implies wall-clock dependence.
+    pub dependence: Vec<Self>,
+}
+
+impl TimeDependence {
+    /// Construct a new [`TimeDependence`] from an optional schedule and a collection of
+    /// dependencies.
+    pub fn new(schedule: Option<RefreshSchedule>, dependence: Vec<Self>) -> Self {
+        Self {
+            schedule,
+            dependence,
+        }
+    }
+
+    /// Normalizes by removing unnecessary nesting.
+    pub fn normalize(this: Option<Self>) -> Option<Self> {
+        match this {
+            Some(TimeDependence {
+                schedule: None,
+                mut dependence,
+            }) if dependence.len() == 1 => Some(dependence.remove(0)),
+            other => other,
+        }
+    }
+
+    /// Applies the indefiniteness to a wall clock time.
+    pub fn apply(&self, wall_clock: Timestamp) -> Option<Timestamp> {
+        let result = self
+            .dependence
+            .iter()
+            .map(|inner| inner.apply(wall_clock))
+            .min()
+            .unwrap_or(Some(wall_clock))?;
+
+        if let Some(schedule) = &self.schedule {
+            schedule.round_up_timestamp(result)
+        } else {
+            Some(result)
+        }
+    }
+}
+
+impl RustType<ProtoTimeDependence> for TimeDependence {
+    fn into_proto(&self) -> ProtoTimeDependence {
+        ProtoTimeDependence {
+            schedule: self.schedule.as_ref().map(|s| s.into_proto()),
+            dependence: self.dependence.into_proto(),
+        }
+    }
+
+    fn from_proto(proto: ProtoTimeDependence) -> Result<Self, TryFromProtoError> {
+        Ok(TimeDependence {
+            schedule: proto
+                .schedule
+                .map(RefreshSchedule::from_proto)
+                .transpose()?,
+            dependence: proto
+                .dependence
+                .into_iter()
+                .map(TimeDependence::from_proto)
+                .collect::<Result<_, _>>()?,
+        })
+    }
+}
+
+impl Arbitrary for TimeDependence {
+    type Strategy = BoxedStrategy<Self>;
+    type Parameters = ();
+
+    fn arbitrary_with(_args: Self::Parameters) -> Self::Strategy {
+        any::<Option<RefreshSchedule>>()
+            .prop_map(|s| TimeDependence::new(s, vec![]))
+            .boxed()
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    #[mz_ore::test]
+    fn test_time_dependence_normalize() {
```

Comment:
In both of the tests in `test_time_dependence_normalize`, `normalize` doesn't have any actual work to do, i.e., the struct is unchanged. (But `normalize` is pretty simple, so it's not a big problem.)